lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp

```diff
@@ -0,0 +1,293 @@
+          // we had a bitsize but no value, so move the offset forward I guess.
+          offset += (bitsize / 8);
+          return false;
+        }
+        ent->GetAsDictionary()->AddIntegerItem("offset", offset);
+        Status error;
+        switch (bitsize / 8) {
+        case 2: {
+          Scalar value_scalar((uint16_t)value);
+          value_scalar.GetAsMemoryData(m_register_data.data() + offset, 2,
+                                       byte_order, error);
+          offset += 2;
+        } break;
+        case 4: {
+          Scalar value_scalar((uint32_t)value);
+          value_scalar.GetAsMemoryData(m_register_data.data() + offset, 4,
+                                       byte_order, error);
+          offset += 4;
+        } break;
+        case 8: {
+          Scalar value_scalar((uint64_t)value);
+          value_scalar.GetAsMemoryData(m_register_data.data() + offset, 8,
+                                       byte_order, error);
+          offset += 8;
+        } break;
+        }
+        return true;
+      });
+  }
+
+  // Create a DynamicRegisterInfo from the metadata JSON.
+  std::unique_ptr<DynamicRegisterInfo> additional_reginfo_up;
+  if (metadata_registers_dict)
+    additional_reginfo_up = DynamicRegisterInfo::Create(
+        *metadata_registers_dict, target.GetArchitecture());
+
+  // Copy the core thread register sets into our combined register set list.
+  for (size_t idx = 0; idx < core_thread_regctx_sp->GetRegisterSetCount();
+       idx++) {
+    RegisterSet new_set;
+    const RegisterSet *old_set = core_thread_regctx_sp->GetRegisterSet(idx);
+    new_set.name = ConstString(old_set->name).AsCString();
+    if (old_set->short_name)
+      new_set.short_name = ConstString(old_set->short_name).AsCString();
+    else
+      new_set.short_name = nullptr;
+    m_register_sets.push_back(new_set);
+  }
+
+  // Set up our RegisterSet array.
+  // First copying all of the core thread register sets,
+  // then any additional unique register sets from the metadata.
+  if (additional_reginfo_up) {
+    for (size_t idx = 0; idx < additional_reginfo_up->GetNumRegisterSets();
+         idx++) {
+      bool found_match = false;
+      const RegisterSet *old_set = additional_reginfo_up->GetRegisterSet(idx);
+      for (size_t jdx = 0; jdx < m_register_sets.size(); jdx++) {
+        if (strcmp(m_register_sets[jdx].name, old_set->name) == 0) {
+          metadata_regset_to_combined_regset[idx] = jdx;
+          found_match = true;
+          break;
+        }
+      }
+      if (!found_match) {
+        RegisterSet new_set;
+        new_set.name = ConstString(old_set->name).AsCString();
+        if (old_set->short_name)
+          new_set.short_name = ConstString(old_set->short_name).AsCString();
+        else
+          new_set.short_name = nullptr;
+        metadata_regset_to_combined_regset[idx] = m_register_sets.size();
+        m_register_sets.push_back(new_set);
+      }
+    }
+  }
```

Comment:
This seems almost (?) identical to the code above? Can this be extracted into a helper that's passed `metadata_registers_dict` and `additional_reginfo_up` respectively?